### PR TITLE
Prepare to make StringContext.{f,raw} into macros

### PR DESCRIPTION
**Pull request by adriaanm**
hijack string interpolators during typer

after bootstrap, we can turn the def s / def raw's RHS into macro ???
until then, pretend they are

Fix https://github.com/scala/bug/issues/11366
Backstop for scala/bug#10870, but could be improved

### CODE REVIEW

**Comment by smarter** [comment]
> after bootstrap, we can turn the def s / def raw's RHS into macro ???

That will break any use of these methods from projects compiled with Dotty. We might be able to shadow them to avoid that but we'll need to coordinate on that.

**Comment by adriaanm** [comment]
We'd like to perform the bootstrap before RC1. Do you think that's workable?

**Comment by smarter** [comment]
Not sure, we still have some work to do to catch up with the recent changes to the standard library (unapply for string interpolators, ...).

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 1]
```diff
@@ -1578,6 +1578,10 @@ trait Definitions extends api.StandardDefinitions {
 
     /** Efficient access to member symbols which must be looked up each run. Access via `currentRun.runDefinitions` */
     final class RunDefinitions {
+      // until we bootstrap and they are implemented as `macro ???` to trigger fasttrack directly
+      def treatLikeMacro(sym: Symbol) =
+        sym.isMethod && ((sym.name == nme.s || sym.name == nme.raw_) && sym.owner == StringContextClass) && (sym == StringContext_s || sym == StringContext_raw)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala:1583]: Why not just:

```
def treatLikeMacro(sym: Symbol) = (sym eq StringContext_s) || (sym eq StringContext_raw)
```

**Comment by retronym** [comment]
This is a novel approach to fast-tracking the way to FastTrack. I think I understand it, seems okay :)

Have you locally tried the followup change? Could you share that branch so we can see the end result?

File: src/compiler/scala/tools/reflect/FastStringInterpolator.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      parts.length == (args.length + 1) &&
+      args.length <= 64 => // TODO make more robust to large input so that we can drop this condition, chunk the concatenations in manageable batches
+
+      val treated =
+        if (isRaw) parts
+        else
+          try
+            parts.mapConserve { case lit@Literal(Constant(stringVal: String)) =>
+              val k = Constant(StringContext.processEscapes(stringVal))
+              // To avoid the backlash of backslash, taken literally by Literal, escapes are processed strictly (scala/bug#11196)
```
Comment (som-snytt) [inline src/compiler/scala/tools/reflect/FastStringInterpolator.scala:37]: That's quite some comment archaeology.

**Review by retronym — APPROVED** [review]

File: src/reflect/scala/reflect/internal/Definitions.scala (hunk 1 above)
Comment (adriaanm) [reply src/reflect/scala/reflect/internal/Definitions.scala:1583]: This method seemed to be hit a lot.

Comment (adriaanm) [reply src/reflect/scala/reflect/internal/Definitions.scala:1583]: (Early on in type checking, so didn’t want to force symbols when not needed.)

**Comment by adriaanm** [comment]
Yep, I’ve confirmed that STARRing this PR, the method bodies of the two interpolators can be `macro ???`. That’s the entirety of the patch (although we can also remove the ugly treatLikeMacro)

**Comment by SethTisue** [comment]
we have now re-STARRed

**Pull request merged**